File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    arr = [[0] * n for _ in range(n)]
    
    dr = [0, 1, 0, -1]
    dc = [1, 0, -1, 0]
    
    r, c = 0, 0
    d = 0
    num = 1
    
    while num <= n * n:
        arr[r][c] = num
        num += 1
        
        nr, nc = r + dr[d], c + dc[d]
        
        if not (0 <= nr < n and 0 <= nc < n and arr[nr][nc] == 0):
            d = (d + 1) % 4
            nr, nc = r + dr[d], c + dc[d]
            
        r, c = nr, nc
        
    return arr

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.js
function solution(n) {
    const arr = Array.from({ length: n }, () => Array(n).fill(0));
    
    const dr = [0, 1, 0, -1];
    const dc = [1, 0, -1, 0];

    let r = 0;
    let c = 0;
    let d = 0;
    let num = 1;
    
    while(num <= n * n) {
        arr[r][c] = num;
        num += 1;

        let nr = r + dr[d];
        let nc = c + dc[d];

        if (!(nr >= 0 && nr < n && nc >= 0 && nc < n && arr[nr][nc] === 0)) {
            d = (d + 1) % 4;
            nr = r + dr[d];
            nc = c + dc[d];
        }
        
        r = nr;
        c = nc;
    }
    
    return arr;
}